### PR TITLE
Add proton internal consumer

### PR DESCRIPTION
**Pull request by azarakovskiy**
Adds a bit of smartness to Proton and allows to skip the usage of Kafkacat completely when using it. Supports some things to format the output or specify the offset.

Usage:
```
proton consume \
  -b kafka.broker \
  -t my-topic \
  -m ./schema.proto \
  -o s@1646218000000 -o e@1646218099197 \
  -f "%T \t %k\t%s"
  --key "my-key-0123\d*"
  --verbose
```

### CODE REVIEW

File: internal/consumer/consumer.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	config := sarama.NewConfig()
+	config.ClientID = "proton-consumer"
+	config.Consumer.Return.Errors = true
+	config.Version = sarama.V0_11_0_0
+	config.Consumer.IsolationLevel = sarama.ReadCommitted
+
+	fmt.Println("Spinning the wheel... Connecting, gathering partitions data and stuff...")
+
+	url := cfg.URL
+	if !strings.HasSuffix(url, ":9092") {
```
Comment (wimspaargaren) [inline internal/consumer/consumer.go:56]: Do we need to make this configurable? Also, what if (for some reason) I'm running my broker at <domain>:8080, will this result in `<domain>:8080:9092` ?

File: internal/consumer/consumer.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+								Value: msg,
+								Topic: topic,
+								Time:  message.Timestamp,
+							})
+						} else {
+							k.printer.PrintErr(err)
+						}
+
+						if o.end == message.Offset {
+							k.log(fmt.Sprintf("# Reached stop timestamp for topic %s: exiting", offsetMsg(topic, o.partition, o.end)))
```
Comment (wimspaargaren) [inline internal/consumer/consumer.go:168]: I was thinking, maybe we should not stop here and let the user decide to cancel consuming(with kill which is handled in Run). WDYT?

**Comment by wimspaargaren** [comment]
Very good job Alexey! 🚀

File: cmd/consume.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+var topic string
+var broker string
+var proto string
+var format string
+var keyGrep string
+var offsets []string
+var startTime, endTime int64
+var verbose bool
+
+func init() {
```
[the remark this answers is not in the document]
Comment (azarakovskiy) [reply cmd/consume.go:35]: Ehm, I am not sure why and whether it's needed. I copied it from another implementation. 
Let me take a quick look, maybe I could improve some stuff.

[the remark this answers is not in the document]
Comment (azarakovskiy) [reply cmd/consume.go:36]: Cobra is a weird tool, I must confess.

I couldn't get rid if `init()` and make it work with a function like `getConfig()`. It just silently stops working. 

However, I could improve a couple of things. https://github.com/beatlabs/proton/pull/13/commits/69a5d7b4852693bf4073c518df0e30ca1e19635f

File: internal/consumer/consumer.go (hunk 1 above)
Comment (azarakovskiy) [reply internal/consumer/consumer.go]: That's what kafkacat does as well, but I made it slightly smarter https://github.com/beatlabs/proton/pull/13/commits/2c69c9ef18dcb5f127a622b48aad685aeef76643

File: internal/consumer/consumer.go (hunk 2 above)
Comment (azarakovskiy) [reply internal/consumer/consumer.go:178]: if we reach the end of the requested offset, we will never have any more messages. If the end offset is not specified, the tool won't quit. here: https://github.com/beatlabs/proton/pull/13/files/266f22793c5ef9e743de40b355386c75f263a69b#diff-9fa0ff7220984702656a024de0f80ac1a44878f5ba3b2ba097eb0f20514ac305R172

**Review by wimspaargaren — APPROVED** [review]

**Review by wimspaargaren — APPROVED** [review]

File: internal/consumer/consumer.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+					errCh <- err
+					return
+				}
+
+				for {
+					select {
+					case <-k.ctx.Done():
+						return
+					case message := <-c.Messages():
+						if k.keyGrep.Match(message.Key) {
```
Comment (peterklijn) [inline internal/consumer/consumer.go:162]: This code block is quite big and has a high cyclomatic complexity, perhaps you can extract the message processing to its own function?

File: internal/output/printer.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	val = strings.ReplaceAll(val, "\\t", "\t")
+	val = strings.ReplaceAll(val, "\\n", "\n")
+	val = strings.ReplaceAll(val, "\\r", "\r")
+	val = strings.ReplaceAll(val, "%s", msg.Value)
+	val = strings.ReplaceAll(val, "%k", msg.Key)
+	val = strings.ReplaceAll(val, "%t", msg.Topic)
+	val = strings.ReplaceAll(val, "%p", fmt.Sprintf("%d", msg.Partition))
+	val = strings.ReplaceAll(val, "%o", fmt.Sprintf("%d", msg.Offset))
+	val = strings.ReplaceAll(val, "%Tf", msg.Time.Format(time.RFC3339))
+	val = strings.ReplaceAll(val, "%T", strconv.Itoa(int(msg.Time.UnixMilli())))
```
Comment (peterklijn) [inline internal/output/printer.go:63]: If this works already, I think your comments [here](https://github.com/beatlabs/proton/pull/13/files#diff-bebec131d68669c9489ca25a38fbd21dd383eb7bdc47ded64288323fd75d15c8R66-R72) are outdated

**Review by peterklijn — COMMENTED** [review]
Nice work! 
A few comments :)

File: cmd/consume.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	// [not yet supported] %K                 Message key length (or -1 for NULL)
+	// [not yet supported] %h                 Message headers (n=v CSV)	
+	// [not yet supported] %p                 Partition
+	// [not yet supported] %o                 Message offset	
+Example:
+	-f 'Key: %k, Time: %Tf \nValue: %s'`)
+
+	// FIXME: kafkacat's syntax allows specifying offsets using an array of `-o` flags.
+	// Specifying `-o 123 -o 234` doesn't work with Cobra for an unknown reason but it actually should.
+	// So before it's fixed, using the non-conventional `-s 123456789` and `-e 234567890`. It should be `-o s@123456789 -o e@234567890` instead.
```
[the remark this answers is not in the document]
Comment (azarakovskiy) [reply cmd/consume.go]: I want it to be as close as possible to kcat syntax-wise. So I just fixed stuff that didn't work.
https://github.com/beatlabs/proton/pull/13/commits/ba90f93e70fae9aeaf3160df96f92ff991eaf220

File: internal/output/printer.go (hunk 5 above)
Comment (azarakovskiy) [reply internal/output/printer.go:63]: Not sure what comments you're talking about.

Comment (peterklijn) [reply internal/output/printer.go:63]: You already fixed it :)

**Review by peterklijn — APPROVED** [review]
Nice work!

**Pull request merged**